File: clients/ts/src/ingest.ts
import * as z from 'zod';

import type {
  OptoSyncClient,
  QueueBatchMutation,
} from './client.js';
import {
  OptimismLevel,
  executeReactiveWrite,
  type ReactiveWriteResult,
} from './reactive.js';

export const SYNC_INGEST_FORMAT = 'opto-sync.ingest.v1' as const;
export const SYNC_INGEST_SCHEMA_ID =
  'https://opto-sync.dev/schemas/opto-sync-ingest.v1.schema.json';
export const DEFAULT_MAX_INGEST_BYTES = 16 * 1024 * 1024;

const scopeIdSchema = z
  .string()
  .min(1)
  .max(128)
  .regex(/^[A-Za-z0-9][A-Za-z0-9._:-]*$/);
const recordIdSchema = z.string().min(1).max(512);
const canonicalRevisionSchema = z.string().regex(/^(0|[1-9][0-9]*)$/);
const hlcTimestampSchema = z
  .string()
  .regex(/^[0-9]{13}-[0-9a-f]{4}-[A-Za-z0-9._:]+$/);
const decimalTimestampSchema = z.string().regex(/^(0|[1-9][0-9]*)$/);

/**
 * The source JSON Schema deliberately rejects JSON numbers for timestamps.
 * A decimal string survives JavaScript, Dart, Rust, Gleam, SQLite JSON, and
 * PostgreSQL JSONB without losing a 64-bit nanosecond value.
 */
export const syncTimestampSchema = z.union([
  hlcTimestampSchema,
  z.iso.datetime({ offset: true }),
  decimalTimestampSchema,
]);

export const syncIngestRecordSchema = z
  .object({
    updatedAt: syncTimestampSchema,
    createdAt: syncTimestampSchema.optional(),
    syncedAt: syncTimestampSchema.optional(),
  })
  .catchall(z.json());

const syncIngestUpsertSchema = z.strictObject({
  operation: z.literal('upsert'),
  table: scopeIdSchema,
  recordId: recordIdSchema,
  record: syncIngestRecordSchema,
  baseRevision: canonicalRevisionSchema.optional(),
  resurrect: z.boolean().optional(),
});

const syncIngestDeleteSchema = z.strictObject({
  operation: z.literal('delete'),
  table: scopeIdSchema,
  recordId: recordIdSchema,
  deletedAt: syncTimestampSchema,
  baseRevision: canonicalRevisionSchema.optional(),
});

export const syncIngestMutationSchema = z.discriminatedUnion('operation', [
  syncIngestUpsertSchema,
  syncIngestDeleteSchema,
]);

/**
 * Runtime mirror of `schemas/opto-sync-ingest.v1.schema.json`.
 *
 * The checked-in JSON Schema is the cross-language source of truth. Shared
 * valid/invalid fixtures are exercised against this Zod schema and the Dart,
 * Rust, and Gleam decoders to prevent validator drift.
 */
export const syncIngestDocumentSchema = z.strictObject({
  format: z.literal(SYNC_INGEST_FORMAT),
  batchId: scopeIdSchema,
  createdAt: syncTimestampSchema,
  mutations: z.array(syncIngestMutationSchema).min(1).max(10_000),
});

export type SyncTimestamp = z.infer<typeof syncTimestampSchema>;
export type SyncIngestRecord = z.infer<typeof syncIngestRecordSchema>;
export type SyncIngestMutation = z.infer<typeof syncIngestMutationSchema>;
export type SyncIngestDocument = z.infer<typeof syncIngestDocumentSchema>;

export type SyncIngestInput =
  | unknown
  | string
  | Blob
  | ArrayBuffer
  | ArrayBufferView;

export interface ParseSyncIngestOptions {
  maxBytes?: number;
}

export class SyncIngestValidationError extends Error {
  readonly issues: readonly {
    path: readonly PropertyKey[];
    message: string;
  }[];

  constructor(
    message: string,
    issues: readonly {
      path: readonly PropertyKey[];
      message: string;
    }[] = [],
  ) {
    super(message);
    this.name = 'SyncIngestValidationError';
    this.issues = issues;
  }
}

function assertMaxBytes(bytes: number, maxBytes: number): void {
  if (bytes > maxBytes) {
    throw new SyncIngestValidationError(
      `ingest document is ${bytes} bytes; limit is ${maxBytes}`,
    );
  }
}

function decodeUtf8(bytes: Uint8Array): string {
  try {
    return new TextDecoder('utf-8', { fatal: true }).decode(bytes);
  } catch {
    throw new SyncIngestValidationError(
      'ingest document is not valid UTF-8',
    );
  }
}

async function normalizeInput(
  input: SyncIngestInput,
  maxBytes: number,
): Promise<unknown> {
  if (
    typeof Blob === 'function' &&
    input instanceof Blob
  ) {
    assertMaxBytes(input.size, maxBytes);
    return JSON.parse(decodeUtf8(new Uint8Array(await input.arrayBuffer())));
  }
  if (input instanceof ArrayBuffer) {
    assertMaxBytes(input.byteLength, maxBytes);
    return JSON.parse(decodeUtf8(new Uint8Array(input)));
  }
  if (ArrayBuffer.isView(input)) {
    assertMaxBytes(input.byteLength, maxBytes);
    return JSON.parse(
      decodeUtf8(
        new Uint8Array(input.buffer, input.byteOffset, input.byteLength),
      ),
    );
  }
  if (typeof input === 'string') {
    assertMaxBytes(new TextEncoder().encode(input).byteLength, maxBytes);
    return JSON.parse(input);
  }
  return input;
}

/**
 * Parse a JSON string, browser Blob/File, byte buffer, or already-decoded
 * object without writing to local storage.
 */
export async function parseSyncIngestDocument(
  input: SyncIngestInput,
  options: ParseSyncIngestOptions = {},
): Promise<SyncIngestDocument> {
  const maxBytes = options.maxBytes ?? DEFAULT_MAX_INGEST_BYTES;
  if (!Number.isSafeInteger(maxBytes) || maxBytes < 1) {
    throw new RangeError('maxBytes must be a positive safe integer');
  }

  let decoded: unknown;
  try {
    decoded = await normalizeInput(input, maxBytes);
  } catch (error) {
    if (error instanceof SyncIngestValidationError) throw error;
    throw new SyncIngestValidationError('ingest document is not valid JSON');
  }
  const result = syncIngestDocumentSchema.safeParse(decoded);
  if (!result.success) {
    throw new SyncIngestValidationError(
      'ingest document does not match opto-sync.ingest.v1',
      result.error.issues.map((issue) => ({
        path: issue.path,
        message: issue.message,
      })),
    );
  }
  return result.data;
}

function toQueueMutation(
  mutation: SyncIngestMutation,
): QueueBatchMutation {
  if (mutation.operation === 'delete') {
    return {
      operation: 'delete',
      tableName: mutation.table,
      recordId: mutation.recordId,
      baseRevision: mutation.baseRevision,
    };
  }
  return {
    operation: 'upsert',
    tableName: mutation.table,
    recordId: mutation.recordId,
    payload: mutation.record,
    baseRevision: mutation.baseRevision,
    resurrect: mutation.resurrect,
  };
}

export interface IngestSyncDocumentOptions<RemoteResult> {
  input: SyncIngestInput;
  client: Pick<OptoSyncClient, 'queueBatch'>;
  optimism?: OptimismLevel;
  maxBytes?: number;
  /**
   * Required only for `ServerConfirmed`. Applications can POST the validated
   * document to their own authenticated ingest endpoint.
   */
  remoteIngest?: (
    document: Readonly<SyncIngestDocument>,
  ) => Promise<RemoteResult>;
  installRemote?: (result: RemoteResult) => Promise<void>;
  requestBackgroundSync?: () => void | Promise<void>;
  syncNow?: () => Promise<void>;
}

export interface IngestSyncDocumentResult<RemoteResult> {
  document: SyncIngestDocument;
  write: ReactiveWriteResult<RemoteResult, readonly number[]>;
}

/**
 * Validate the complete document, then apply the selected write strategy.
 *
 * Durable strategies use one atomic IndexedDB batch and leave protocol v1 to
 * carry the mutations to PostgreSQL/Supabase. Server-confirmed mode performs
 * no local queue write and requires an application-owned authenticated API.
 */
export async function ingestSyncDocument<RemoteResult = unknown>(
  options: IngestSyncDocumentOptions<RemoteResult>,
): Promise<IngestSyncDocumentResult<RemoteResult>> {
  const document = await parseSyncIngestDocument(options.input, {
    maxBytes: options.maxBytes,
  });
  const optimism = options.optimism ?? OptimismLevel.DurableLocal;
  const write = await executeReactiveWrite<
    RemoteResult,
    readonly number[]
  >({
    optimism,
    remoteWrite: async () => {
      if (!options.remoteIngest) {
        throw new TypeError(
          'server-confirmed ingest requires remoteIngest',
        );
      }
      return options.remoteIngest(document);
    },
    queueLocal: () =>
      options.client.queueBatch(document.mutations.map(toQueueMutation)),
    installRemote: options.installRemote,
    requestBackgroundSync: options.requestBackgroundSync,
    syncNow: options.syncNow,
  });
  return { document, write };
}

File: clients/dart/lib/src/ingest.dart
import 'dart:async';
import 'dart:convert';
import 'dart:typed_data';

import '../opto_sync_client.dart'
    show
        OptoSyncClient,
        QueueBatchDelete,
        QueueBatchMutation,
        QueueBatchUpsert;
import 'reactive_sync.dart';

const String syncIngestFormat = 'opto-sync.ingest.v1';
const String syncIngestSchemaId =
    'https://opto-sync.dev/schemas/opto-sync-ingest.v1.schema.json';
const int defaultMaxIngestBytes = 16 * 1024 * 1024;

final RegExp _scopeId = RegExp(r'^[A-Za-z0-9][A-Za-z0-9._:-]*$');
final RegExp _canonicalDecimal = RegExp(r'^(0|[1-9][0-9]*)$');
final RegExp _hlc = RegExp(
  r'^[0-9]{13}-[0-9a-f]{4}-[A-Za-z0-9._:]+$',
);
final RegExp _rfc3339 = RegExp(
  r'^[0-9]{4}-[0-9]{2}-[0-9]{2}T'
  r'[0-9]{2}:[0-9]{2}:[0-9]{2}'
  r'(?:\.[0-9]+)?(?:Z|[+-][0-9]{2}:[0-9]{2})$',
);

final class SyncIngestValidationException implements Exception {
  final String path;
  final String message;

  const SyncIngestValidationException(this.path, this.message);

  @override
  String toString() =>
      'SyncIngestValidationException($path): $message';
}

sealed class SyncIngestMutation {
  final String table;
  final String recordId;
  final String? baseRevision;

  const SyncIngestMutation({
    required this.table,
    required this.recordId,
    this.baseRevision,
  });
}

final class SyncIngestUpsert extends SyncIngestMutation {
  final Map<String, dynamic> record;
  final bool resurrect;

  const SyncIngestUpsert({
    required super.table,
    required super.recordId,
    required this.record,
    super.baseRevision,
    this.resurrect = false,
  });
}

final class SyncIngestDelete extends SyncIngestMutation {
  final String deletedAt;

  const SyncIngestDelete({
    required super.table,
    required super.recordId,
    required this.deletedAt,
    super.baseRevision,
  });
}

final class SyncIngestDocument {
  final String format;
  final String batchId;
  final String createdAt;
  final List<SyncIngestMutation> mutations;

  const SyncIngestDocument({
    required this.format,
    required this.batchId,
    required this.createdAt,
    required this.mutations,
  });
}

Never _invalid(String path, String message) =>
    throw SyncIngestValidationException(path, message);

bool _isSyncTimestamp(Object? value) {
  if (value is! String) return false;
  if (_hlc.hasMatch(value) || _canonicalDecimal.hasMatch(value)) return true;
  return _rfc3339.hasMatch(value) && DateTime.tryParse(value) != null;
}

void _requireExactKeys(
  Map<String, dynamic> value,
  Set<String> required,
  Set<String> allowed,
  String path,
) {
  for (final key in required) {
    if (!value.containsKey(key)) _invalid(path, 'missing required key $key');
  }
  for (final key in value.keys) {
    if (!allowed.contains(key)) _invalid(path, 'unknown key $key');
  }
}

String _requireScopeId(Object? value, String path) {
  if (value is! String ||
      value.isEmpty ||
      value.length > 128 ||
      !_scopeId.hasMatch(value)) {
    _invalid(path, 'must be a valid 1-128 character scope id');
  }
  return value;
}

String _requireRecordId(Object? value, String path) {
  if (value is! String || value.isEmpty || value.length > 512) {
    _invalid(path, 'must be a non-empty string of at most 512 characters');
  }
  return value;
}

String? _optionalRevision(Object? value, String path) {
  if (value == null) return null;
  if (value is! String || !_canonicalDecimal.hasMatch(value)) {
    _invalid(path, 'must be a canonical unsigned decimal string');
  }
  return value;
}

Map<String, dynamic> _requireRecord(Object? value, String path) {
  if (value is! Map<String, dynamic>) {
    _invalid(path, 'must be a JSON object');
  }
  if (!_isSyncTimestamp(value['updatedAt'])) {
    _invalid('$path.updatedAt', 'must be a sync timestamp string');
  }
  for (final key in ['createdAt', 'syncedAt']) {
    if (value.containsKey(key) && !_isSyncTimestamp(value[key])) {
      _invalid('$path.$key', 'must be a sync timestamp string');
    }
  }
  return value;
}

SyncIngestMutation _parseMutation(
  Object? value,
  int index,
) {
  final path = r'$.mutations[' '$index]';
  if (value is! Map<String, dynamic>) {
    _invalid(path, 'must be a JSON object');
  }
  final operation = value['operation'];
  final table = _requireScopeId(value['table'], '$path.table');
  final recordId = _requireRecordId(value['recordId'], '$path.recordId');
  final baseRevision = _optionalRevision(
    value['baseRevision'],
    '$path.baseRevision',
  );
  if (operation == 'upsert') {
    _requireExactKeys(
      value,
      {'operation', 'table', 'recordId', 'record'},
      {
        'operation',
        'table',
        'recordId',
        'record',
        'baseRevision',
        'resurrect',
      },
      path,
    );
    final resurrect = value['resurrect'];
    if (resurrect != null && resurrect is! bool) {
      _invalid('$path.resurrect', 'must be a boolean');
    }
    return SyncIngestUpsert(
      table: table,
      recordId: recordId,
      record: _requireRecord(value['record'], '$path.record'),
      baseRevision: baseRevision,
      resurrect: resurrect as bool? ?? false,
    );
  }
  if (operation == 'delete') {
    _requireExactKeys(
      value,
      {'operation', 'table', 'recordId', 'deletedAt'},
      {'operation', 'table', 'recordId', 'deletedAt', 'baseRevision'},
      path,
    );
    final deletedAt = value['deletedAt'];
    if (!_isSyncTimestamp(deletedAt)) {
      _invalid('$path.deletedAt', 'must be a sync timestamp string');
    }
    return SyncIngestDelete(
      table: table,
      recordId: recordId,
      deletedAt: deletedAt as String,
      baseRevision: baseRevision,
    );
  }
  _invalid('$path.operation', 'must be upsert or delete');
}

Object? _decodeInput(Object? input, int maxBytes) {
  String encoded;
  if (input is String) {
    encoded = input;
  } else if (input is ByteBuffer) {
    encoded = utf8.decode(input.asUint8List(), allowMalformed: false);
  } else if (input is List<int>) {
    encoded = utf8.decode(input, allowMalformed: false);
  } else {
    try {
      encoded = jsonEncode(input);
    } on Object {
      _invalid(r'$', 'must contain only JSON values');
    }
  }
  final bytes = utf8.encode(encoded).length;
  if (bytes > maxBytes) {
    _invalid(r'$', 'document is $bytes bytes; limit is $maxBytes');
  }
  try {
    return jsonDecode(encoded);
  } on FormatException {
    _invalid(r'$', 'is not valid JSON');
  }
}

/// Parse and validate an object, JSON string, file bytes, or browser blob bytes
/// against `schemas/opto-sync-ingest.v1.schema.json`.
Future<SyncIngestDocument> parseSyncIngestDocument(
  Object? input, {
  int maxBytes = defaultMaxIngestBytes,
}) async {
  if (maxBytes < 1) {
    throw RangeError.range(maxBytes, 1, null, 'maxBytes');
  }
  final decoded = _decodeInput(input, maxBytes);
  if (decoded is! Map<String, dynamic>) {
    _invalid(r'$', 'must be a JSON object');
  }
  _requireExactKeys(
    decoded,
    {'format', 'batchId', 'createdAt', 'mutations'},
    {'format', 'batchId', 'createdAt', 'mutations'},
    r'$',
  );
  if (decoded['format'] != syncIngestFormat) {
    _invalid(r'$.format', 'must equal $syncIngestFormat');
  }
  final batchId = _requireScopeId(decoded['batchId'], r'$.batchId');
  final createdAt = decoded['createdAt'];
  if (!_isSyncTimestamp(createdAt)) {
    _invalid(r'$.createdAt', 'must be a sync timestamp string');
  }
  final mutations = decoded['mutations'];
  if (mutations is! List ||
      mutations.isEmpty ||
      mutations.length > 10000) {
    _invalid(r'$.mutations', 'must contain 1-10000 mutations');
  }
  return SyncIngestDocument(
    format: syncIngestFormat,
    batchId: batchId,
    createdAt: createdAt as String,
    mutations: [
      for (var index = 0; index < mutations.length; index++)
        _parseMutation(mutations[index], index),
    ],
  );
}

QueueBatchMutation _toQueueMutation(SyncIngestMutation mutation) {
  if (mutation is SyncIngestDelete) {
    return QueueBatchDelete(
      tableName: mutation.table,
      recordId: mutation.recordId,
      baseRevision: mutation.baseRevision,
    );
  }
  final upsert = mutation as SyncIngestUpsert;
  return QueueBatchUpsert(
    tableName: upsert.table,
    recordId: upsert.recordId,
    payload: upsert.record,
    baseRevision: upsert.baseRevision,
    resurrect: upsert.resurrect,
  );
}

final class IngestSyncDocumentResult {
  final SyncIngestDocument document;
  final ReactiveWriteResult<Object?, List<int>> write;

  const IngestSyncDocumentResult({
    required this.document,
    required this.write,
  });
}

/// Validate the complete document, then apply one explicit optimism strategy.
Future<IngestSyncDocumentResult> ingestSyncDocument({
  required Object? input,
  required OptoSyncClient client,
  OptimismLevel optimism = OptimismLevel.durableLocal,
  int maxBytes = defaultMaxIngestBytes,
  Future<Object?> Function(SyncIngestDocument document)? remoteIngest,
  Future<void> Function(Object? result)? installRemote,
  FutureOr<void> Function()? requestBackgroundSync,
  Future<void> Function()? syncNow,
}) async {
  final document = await parseSyncIngestDocument(
    input,
    maxBytes: maxBytes,
  );
  final write = await executeReactiveWrite<Object?, List<int>>(
    optimism: optimism,
    remoteWrite: () {
      if (remoteIngest == null) {
        throw ArgumentError('server-confirmed ingest requires remoteIngest');
      }
      return remoteIngest(document);
    },
    queueLocal: () =>
        client.queueBatch(document.mutations.map(_toQueueMutation).toList()),
    installRemote: installRemote,
    requestBackgroundSync: requestBackgroundSync,
    syncNow: syncNow,
  );
  return IngestSyncDocumentResult(document: document, write: write);
}
